File: sdk/python/signet_client.py
# Copyright 2025 ODIN Protocol Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Signet Protocol - Simple HTTP Client
One-line helper for verified AI-to-AI communications.
"""

import json
import uuid
import requests
from typing import Dict, Any, Optional, List
from dataclasses import dataclass
from datetime import datetime

@dataclass
class SignetResponse:
    """Response from Signet Protocol exchange."""
    success: bool
    trace_id: str
    normalized: Optional[Dict[str, Any]] = None
    receipt: Optional[Dict[str, Any]] = None
    forwarded: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    status_code: Optional[int] = None

class SignetClient:
    """
    Simple HTTP client for Signet Protocol.
    
    Usage:
        client = SignetClient("http://localhost:8088", "your-api-key")
        result = client.exchange({"amount": 100, "currency": "USD"})
    """
    
    def __init__(
        self,
        signet_url: str,
        api_key: str,
        forward_url: Optional[str] = None,
        tenant: Optional[str] = None,
        timeout: int = 30
    ):
        self.signet_url = signet_url.rstrip('/')
        self.api_key = api_key
        self.forward_url = forward_url
        self.tenant = tenant or "client"
        self.timeout = timeout
        self.session = requests.Session()
        
        # Set default headers
        self.session.headers.update({
            "X-SIGNET-API-Key": self.api_key,
            "Content-Type": "application/json",
            "User-Agent": "signet-client-python/1.0"
        })
    
    def exchange(
        self,
        data: Dict[str, Any],
        forward_url: Optional[str] = None,
        trace_id: Optional[str] = None,
        idempotency_key: Optional[str] = None
    ) -> SignetResponse:
        """
        Send data through Signet Protocol for verification.
        
        Args:
            data: The data to verify (will be converted to invoice format)
            forward_url: Override default forward URL
            trace_id: Specific trace ID (auto-generated if not provided)
            idempotency_key: Idempotency key (auto-generated if not provided)
            
        Returns:
            SignetResponse with verification results
        """
        try:
            # Generate IDs if not provided
            if not trace_id:
                trace_id = f"{self.tenant}-{uuid.uuid4()}"
            if not idempotency_key:
                idempotency_key = f"{trace_id}-{uuid.uuid4()}"
            
            # Convert data to Signet payload format
            payload = self._create_payload(data, trace_id, forward_url)
            
            # Set request headers
            headers = {
                "X-SIGNET-Idempotency-Key": idempotency_key
            }
            
            # Make request
            response = self.session.post(
                f"{self.signet_url}/v1/exchange",
                json=payload,
                headers=headers,
                timeout=self.timeout
            )
            
            # Parse response
            return self._parse_response(response, trace_id)
            
        except requests.RequestException as e:
            return SignetResponse(
                success=False,
                trace_id=trace_id or "unknown",
                error=f"Request failed: {str(e)}"
            )
        except Exception as e:
            return SignetResponse(
                success=False,
                trace_id=trace_id or "unknown",
                error=f"Unexpected error: {str(e)}"
            )
    
    def get_chain(self, trace_id: str) -> Optional[List[Dict[str, Any]]]:
        """
        Retrieve the complete receipt chain for a trace.
        
        Args:
            trace_id: The trace ID to retrieve
            
        Returns:
            List of receipts or None if not found
        """
        try:
            response = self.session.get(
                f"{self.signet_url}/v1/receipts/chain/{trace_id}",
                timeout=self.timeout
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
                
        except Exception:
            return None
    
    def export_chain(self, trace_id: str) -> Optional[Dict[str, Any]]:
        """
        Export a signed receipt chain bundle.
        
        Args:
            trace_id: The trace ID to export
            
        Returns:
            Signed export bundle or None if not found
        """
        try:
            response = self.session.get(
                f"{self.signet_url}/v1/receipts/export/{trace_id}",
                timeout=self.timeout
            )
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
                
        except Exception:
            return None
    
    def health_check(self) -> bool:
        """
        Check if Signet Protocol server is healthy.
        
        Returns:
            True if server is healthy, False otherwise
        """
        try:
            response = self.session.get(
                f"{self.signet_url}/healthz",
                timeout=5
            )
            return response.status_code == 200
        except Exception:
            return False
    
    def _create_payload(
        self,
        data: Dict[str, Any],
        trace_id: str,
        forward_url: Optional[str] = None
    ) -> Dict[str, Any]:
        """Convert user data to Signet Protocol payload format."""
        
        # Use provided forward_url or instance default
        final_forward_url = forward_url or self.forward_url
        
        # Convert data to function arguments JSON
        arguments_json = json.dumps(data)
        
        return {
            "payload_type": "openai.tooluse.invoice.v1",
            "target_type": "invoice.iso20022.v1",
            "trace_id": trace_id,
            "payload": {
                "tool_calls": [{
                    "type": "function",
                    "function": {
                        "name": "create_invoice",
                        "arguments": arguments_json
                    }
                }]
            },
            "forward_url": final_forward_url
        }
    
    def _parse_response(self, response: requests.Response, trace_id: str) -> SignetResponse:
        """Parse HTTP response into SignetResponse."""
        try:
            if response.status_code == 200:
                data = response.json()
                return SignetResponse(
                    success=True,
                    trace_id=data.get("trace_id", trace_id),
                    normalized=data.get("normalized"),
                    receipt=data.get("receipt"),
                    forwarded=data.get("forwarded"),
                    status_code=200
                )
            else:
                # Try to parse error response
                try:
                    error_data = response.json()
                    error_msg = error_data.get("detail", f"HTTP {response.status_code}")
                except:
                    error_msg = f"HTTP {response.status_code}: {response.text[:200]}"
                
                return SignetResponse(
                    success=False,
                    trace_id=trace_id,
                    error=error_msg,
                    status_code=response.status_code
                )
                
        except Exception as e:
            return SignetResponse(
                success=False,
                trace_id=trace_id,
                error=f"Response parsing failed: {str(e)}",
                status_code=response.status_code
            )


# Convenience functions for one-liner usage
def signet_exchange(
    signet_url: str,
    api_key: str,
    data: Dict[str, Any],
    forward_url: Optional[str] = None,
    **kwargs
) -> SignetResponse:
    """
    One-liner to send data through Signet Protocol.
    
    Usage:
        result = signet_exchange(
            "http://localhost:8088",
            "your-api-key", 
            {"amount": 100, "currency": "USD"},
            "https://webhook.site/your-url"
        )
    """
    client = SignetClient(signet_url, api_key, forward_url)
    return client.exchange(data, **kwargs)

def verify_invoice(
    signet_url: str,
    api_key: str,
    invoice_data: Dict[str, Any],
    webhook_url: Optional[str] = None
) -> SignetResponse:
    """
    One-liner to verify invoice data through Signet Protocol.
    
    Usage:
        result = verify_invoice(
            "http://localhost:8088",
            "your-api-key",
            {
                "invoice_id": "INV-001",
                "amount": 1000.00,
                "currency": "USD",
                "customer_name": "Acme Corp",
                "description": "Professional services"
            },
            "https://your-system.com/webhook"
        )
    """
    return signet_exchange(signet_url, api_key, invoice_data, webhook_url)


# Example usage and testing
if __name__ == "__main__":
    # Example usage
    client = SignetClient(
        signet_url="http://localhost:8088",
        api_key="demo_key",
        forward_url="https://postman-echo.com/post"
    )
    
    # Test health check
    if client.health_check():
        print("✅ Signet Protocol server is healthy")
        
        # Test exchange
        invoice_data = {
            "invoice_id": "INV-12345",
            "amount": 1500.00,
            "currency": "USD",
            "customer_name": "Example Customer",
            "description": "Consulting services"
        }
        
        result = client.exchange(invoice_data)
        
        if result.success:
            print(f"✅ Exchange successful!")
            print(f"   Trace ID: {result.trace_id}")
            print(f"   Receipt Hash: {result.receipt.get('receipt_hash') if result.receipt else 'N/A'}")
            print(f"   Forwarded: {'Yes' if result.forwarded else 'No'}")
            
            # Retrieve the chain
            chain = client.get_chain(result.trace_id)
            if chain:
                print(f"   Chain Length: {len(chain)} receipts")
            
            # Export the chain
            export = client.export_chain(result.trace_id)
            if export:
                print(f"   Export Bundle: {len(export.get('chain', []))} receipts")
        else:
            print(f"❌ Exchange failed: {result.error}")
    else:
        print("❌ Signet Protocol server is not available")
    
    # Test one-liner
    print("\n--- One-liner test ---")
    result = verify_invoice(
        "http://localhost:8088",
        "demo_key",
        {
            "invoice_id": "INV-ONELINER",
            "amount": 500.00,
            "currency": "USD",
            "customer_name": "One-liner Customer"
        }
    )
    
    print(f"One-liner result: {'✅ Success' if result.success else '❌ Failed'}")
    if result.error:
        print(f"Error: {result.error}")

File: packages/sdk-js/sdk-js-new/src/signet-client.ts
/**
 * Signet Protocol - JavaScript/TypeScript Client SDK
 * One-line "send + verify" for developers
 */

import { SignetReceipt, SignetResponse, ExchangeOptions } from './types';

export class SignetClient {
  private signetUrl: string;
  private apiKey: string;
  private forwardUrl?: string;
  private tenant: string;
  private timeout: number;
  private autoVerify: boolean;

  constructor(options: {
    signetUrl: string;
    apiKey: string;
    forwardUrl?: string;
    tenant?: string;
    timeout?: number;
    autoVerify?: boolean;
  }) {
    this.signetUrl = options.signetUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.forwardUrl = options.forwardUrl;
    this.tenant = options.tenant || 'client';
    this.timeout = options.timeout || 30000;
    this.autoVerify = options.autoVerify ?? true;
  }

  /**
   * Send data through Signet Protocol for verification
   */
  async exchange(options: Omit<ExchangeOptions, 'apiKey'>): Promise<SignetResponse> {
    try {
      // Generate IDs if not provided
      const traceId = options.traceId || `${this.tenant}-${this.generateUuid()}`;
      const idem = options.idem || `${traceId}-${this.generateUuid()}`;

      // Create payload
      const payload = this.createPayload({
        ...options,
        traceId,
      });

      // Make request
      const response = await this.makeRequest('/v1/exchange', {
        method: 'POST',
        headers: {
          'X-SIGNET-API-Key': this.apiKey,
          'X-SIGNET-Idempotency-Key': idem,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      });

      // Parse response
      const result = await this.parseResponse(response, traceId);

      // Auto-verify receipt if enabled
      if (result.success && result.receipt && this.autoVerify) {
        const { verifyReceipt } = await import('./signet-verify');
        const { valid, reason } = await verifyReceipt(result.receipt);
        if (!valid) {
          return {
            ...result,
            success: false,
            error: `Receipt verification failed: ${reason}`,
          };
        }
      }

      return result;
    } catch (error) {
      return {
        success: false,
        traceId: options.traceId || 'unknown',
        error: `Request failed: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * Retrieve receipt chain for a trace
   */
  async getReceipts(traceId: string): Promise<SignetReceipt[] | null> {
    try {
      const response = await this.makeRequest(`/v1/receipts/chain/${traceId}`, {
        method: 'GET',
        headers: {
          'X-SIGNET-API-Key': this.apiKey,
        },
      });

      if (response.ok) {
        return await response.json();
      }
      return null;
    } catch {
      return null;
    }
  }

  /**
   * Export signed receipt bundle
   */
  async exportBundle(traceId: string): Promise<any | null> {
    try {
      const response = await this.makeRequest(`/v1/receipts/export/${traceId}`, {
        method: 'GET',
        headers: {
          'X-SIGNET-API-Key': this.apiKey,
        },
      });

      if (response.ok) {
        return await response.json();
      }
      return null;
    } catch {
      return null;
    }
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<boolean> {
    try {
      const response = await this.makeRequest('/healthz', {
        method: 'GET',
      });
      return response.ok;
    } catch {
      return false;
    }
  }

  private createPayload(options: Omit<ExchangeOptions, 'apiKey'> & { traceId: string }): any {
    const argumentsJson = JSON.stringify(options.payload);

    return {
      payload_type: options.payloadType || 'openai.tooluse.invoice.v1',
      target_type: options.targetType || 'invoice.iso20022.v1',
      trace_id: options.traceId,
      payload: {
        tool_calls: [{
          type: 'function',
          function: {
            name: 'create_invoice',
            arguments: argumentsJson,
          },
        }],
      },
      forward_url: options.forwardUrl || this.forwardUrl,
    };
  }

  private async makeRequest(path: string, init: RequestInit): Promise<Response> {
    const url = `${this.signetUrl}${path}`;
    
    // Use fetch (browser) or node-fetch (Node.js)
    const fetchFn = this.getFetch();
    
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetchFn(url, {
        ...init,
        signal: controller.signal,
      });
      return response;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async parseResponse(response: Response, traceId: string): Promise<SignetResponse> {
    try {
      if (response.ok) {
        const data = await response.json();
        return {
          success: true,
          traceId: data.trace_id || traceId,
          normalized: data.normalized,
          receipt: data.receipt,
          forwarded: data.forwarded,
          statusCode: response.status,
        };
      } else {
        let errorMsg: string;
        try {
          const errorData = await response.json();
          errorMsg = errorData.detail || `HTTP ${response.status}`;
        } catch {
          errorMsg = `HTTP ${response.status}: ${response.statusText}`;
        }

        return {
          success: false,
          traceId,
          error: errorMsg,
          statusCode: response.status,
        };
      }
    } catch (error) {
      return {
        success: false,
        traceId,
        error: `Response parsing failed: ${error instanceof Error ? error.message : String(error)}`,
        statusCode: response.status,
      };
    }
  }

  private getFetch(): typeof fetch {
    if (typeof fetch !== 'undefined') {
      return fetch;
    }
    
    // Node.js environment - require node-fetch
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const nodeFetch = require('node-fetch');
      return nodeFetch.default || nodeFetch;
    } catch {
      throw new Error('fetch is not available. In Node.js, install node-fetch: npm install node-fetch');
    }
  }

  private generateUuid(): string {
    if (typeof crypto !== 'undefined' && crypto.randomUUID) {
      return crypto.randomUUID();
    }
    
    // Fallback UUID v4 implementation
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }
}

/**
 * One-liner exchange function
 */
export async function exchange(options: ExchangeOptions): Promise<SignetResponse> {
  const client = new SignetClient({
    signetUrl: process.env.SIGNET_URL || 'http://localhost:8088',
    apiKey: options.apiKey,
    forwardUrl: options.forwardUrl,
  });

  return client.exchange(options);
}

/**
 * Convenience function for invoice verification
 */
export async function verifyInvoice(
  apiKey: string,
  invoiceData: any,
  options?: {
    signetUrl?: string;
    forwardUrl?: string;
    autoVerify?: boolean;
  }
): Promise<SignetResponse> {
  return exchange({
    apiKey,
    payload: invoiceData,
    forwardUrl: options?.forwardUrl,
    autoVerify: options?.autoVerify,
  });
}

// Export types
export * from './types';
